README.md
Private use ffmpeg wrapper.

@TODO

- fps set
- audio copy


CHANGELOG.md
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [UNRELEASED]
(nothing yet)

## [2.1.0] - 2020-05-15
### Added
- metadata copied by default
- `--no-meta` switch
- `--overwrite` switch
### Changed
- quality is now 25 by default

## [2.0.0] - 2020-05-15
### Added
- tune is now configurable
### Changed
- tune is now `film` by default
### Dev
- app uses esm now


src/cmd.js
const program = require('commander');
const fs = require("fs");
const createError = require("better-custom-error/dist").default;

const ArgumentError = createError("ArgumentError");

const packageJson = require("../package.json");

const number = (value) => {
    return parseInt(value, 10);
};

const float = (value) => {
    return parseFloat(value);
};

const kbps = (value) => {
    return value + "k";
};

const rotateMap = {
    90: "transpose=1",
    180: ["transpose=2", "transpose=2"],
    270: "transpose=2",
    "-90": "transpose=2",
};

const rotate = (value) => {
    const vf = rotateMap[value];
    if (!vf) {
        throw new ArgumentError(`Rotation ${value} is incorrect`);
    }
    return vf;
};

const filePath = (value) => {
    try {
        fs.accessSync(value);
        return value;
    }
    catch (e) {
        throw new ArgumentError(`File ${value} doesn't exist or is inaccessible.`);
    }
};

module.exports = () => {
    try {
        program.version(packageJson.version, '-v, --version')
            .arguments("<file>")
            .option("-i, --input <path>", "source file to encode", filePath)
            .option(
                "--v:quality <number>",
                "sets video quality (0-51, where 0 is lossless, 17 is visually lossless, 23 is optimal default)",
                number,
                23,
            )
            .option(
                "--v:fps <number>",
                "sets output video framerate",
                float,
            )
            .option(
                "--v:rotate <degrees>",
                "rotates video, only `90`, `180`, `270` are valid values",
                rotate,
            )
            .option(
                "--v:scale <degrees>",
                "scales video, feed it with raw ffmpeg `scale` `vf`, basic example: 1280x720, 1280:-1, -1:720",
            )
            .option(
                "--v:preset <preset>",
                "sets video encoding preset", // @todo list presets as validation
                "slow",
            )
            .option("--a:copy", "copies source audio")
            .option("--a:none", "disables audio")
            .option("--a:channels <number>", "sets audio channels count", number)
            .option(
                "--a:bitrate <number>",
                "sets audio quality (in kbps)",
                kbps,
            )
            .option(
                "--from <string>",
                "cuts video from this time",
            )
            .option(
                "--to <string>",
                "cuts video to this time",
            )
            .parse(process.argv)
        ;
    }
    catch (e) {
        if (e instanceof ArgumentError) {
            console.error("error:", e.message);
            process.exit(1);
        }

        console.error("Unexpected error");
        console.error(e);
        process.exit(1);
    }

    if (!program.input) {
        console.error("error: required option `-i, --input <path>` missing");
        process.exit(1);
    }

    if (program["a:copy"]) {
        if (program["a:bitrate"] || program["a:channels"]) {
            console.error("error: you can't configure audio settings if you set to copy it");
            process.exit(1);
        }
        if (program["a:none"]) {
            console.error("error: you can't disable audio if you set to copy it");
            process.exit(1);
        }
    }

    if (program["a:none"]) {
        if (program["a:bitrate"] || program["a:channels"]) {
            console.error("error: you can't configure audio settings if you disabled");
            process.exit(1);
        }
    }

    return program;
};

src/index.js
#!/usr/bin/env node

const path = require("path");
const { spawn, execFile } = require("child_process");
const fs = require("fs-extra");
const pretty = require("pretty-bytes");
const colors = require("colors/safe");

const cmd = require("./cmd");

const program = cmd();
const inputFileName = path.basename(program.input);
const inputFileNameParts = inputFileName.split(".");
// const inputFileExt = inputFileNameParts.pop();
inputFileNameParts.pop();
const inputFileNameWithoutExt = inputFileNameParts.join(".");
const outputFileName = inputFileNameWithoutExt + ".encoded.mp4";

const makeArray = (value) => {
    return Array.isArray(value) ? value : [value];
};

const encode = () => {
    const vf = [];
    if (program["v:rotate"]) {
        const rotate = makeArray(program["v:rotate"]);
        vf.push(...rotate);
    }

    if (program["v:scale"]) {
        vf.push("scale=" + program["v:scale"]);
    }

    const optionalParams = [];
    if (vf.length) {
        optionalParams.push("-vf", vf.join(","));
    }

    if (program["v:fps"]) {
        optionalParams.push("-r", program["v:fps"]);
    }

    if (program["a:channels"]) {
        optionalParams.push("-ac", program["a:channels"]);
    }

    if (program["a:copy"]) {
        optionalParams.push("-c:a", "copy");
    }
    else if (program["a:none"]) {
        optionalParams.push("-an");
    }
    else {
        optionalParams.push("-c:a", "aac");
    }

    if (program["from"]) {
        optionalParams.push("-ss", program["from"]);
    }

    if (program["to"]) {
        optionalParams.push("-to", program["to"]);
    }

    return new Promise((resolve, reject) => {
        const ff = spawn("ffmpeg", [
            "-n",
            "-i", program.input,
            "-c:v", "libx264",
            "-preset", program["v:preset"],
            "-crf", program["v:quality"],
            ...optionalParams,
            "-vsync", "2",
            "-tune", "zerolatency",
            "-movflags", "+faststart",
            outputFileName,
        ]);
        console.info(colors.green(ff.spawnargs.join(" ")));
        let lastError = "";
        ff.stdout.on("data", data => console.log(String(data)));
        ff.stderr.on("data", data => {
            lastError = String(data);
            console.error(lastError);
        });
        ff.on("close", (code) => {
            console.info("Process exited with code:", code);
            if (!code) {
                resolve();
                return;
            }
            if (lastError.includes("already exists. Exiting")) {
                resolve();
                return;
            }
            reject(new Error("ffmpeg exited with code: " + code));
        });
    });
};

const getModifiedDate = () => {
    return new Promise((resolve, reject) => {
        execFile("date", ["-R", "-r", program.input], (error, stdout, stderr) => {
            if (error) {
                reject(new Error(error)); // @todo can do better error?
                return;
            }
            resolve(stdout);
        })
    });
};

const setModifiedDate = (file, newDate) => {
    return new Promise((resolve, reject) => {
        execFile("touch", ["-d", newDate, file], (error, stdout, stderr) => {
            if (error) {
                reject(new Error(error)); // @todo can do better error?
                return;
            }
            resolve(stdout);
        })
    });
};

const fixDate = async () => {
    const date = await getModifiedDate(program.input);
    await setModifiedDate(outputFileName, date);
};

const sizeDiff = async () => {
    const oldFile = await fs.stat(program.input);
    const newFile = await fs.stat(outputFileName);
    const sizeDiff = oldFile.size - newFile.size;
    console.info("");
    console.info(`Old file size: ${pretty(oldFile.size)}`);
    console.info(`New file size: ${pretty(newFile.size)}`);
    if (sizeDiff > 0) {
        console.info("You have saved", pretty(sizeDiff));
    }
    else if (sizeDiff < 0) {
        console.info("New file is larger by", pretty(-sizeDiff));
    }
    else {
        console.info("File sizes are exactly the same! Amazing!");
    }
};

(async () => {
    try {
        await encode();
        await fixDate();
        await sizeDiff()
        console.info("");;
        console.info("Success!");
    }
    catch (e) {
        console.info("");
        console.error("Error happend");
        console.error(e);
    }
})();


src/index.mjs
#!/usr/bin/env node

import path from "path";
import { spawn, execFile } from "child_process";
import fs from "fs-extra";
import pretty from "pretty-bytes";
import colors from "colors/safe.js";

import cmd from "./cmd.mjs";

const program = cmd();
const inputFileName = path.basename(program.input);
const inputFileNameParts = inputFileName.split(".");
inputFileNameParts.pop();
const inputFileNameWithoutExt = inputFileNameParts.join(".");
const outputFileName = inputFileNameWithoutExt + ".encoded.mp4";

const makeArray = (value) => {
    return Array.isArray(value) ? value : [value];
};

const buildOptionalParams = () => { // eslint-disable-line max-statements
    const vf = [];
    if (program["v:rotate"]) {
        const rotate = makeArray(program["v:rotate"]);
        vf.push(...rotate);
    }

    if (program["v:scale"]) {
        vf.push("scale=" + program["v:scale"]);
    }

    const optionalParams = [];
    if (vf.length) {
        optionalParams.push("-vf", vf.join(","));
    }

    if (program["v:fps"]) {
        optionalParams.push("-r", program["v:fps"]);
    }

    if (program["a:channels"]) {
        optionalParams.push("-ac", program["a:channels"]);
    }

    if (program["a:copy"]) {
        optionalParams.push("-c:a", "copy");
    }
    else if (program["a:none"]) {
        optionalParams.push("-an");
    }
    else {
        optionalParams.push("-c:a", "aac");
    }

    if (program.from) {
        optionalParams.push("-ss", program.from);
    }

    if (program.to) {
        optionalParams.push("-to", program.to);
    }

    return optionalParams;
};

const encode = () => {
    const optionalParams = buildOptionalParams();
    const meta = program.meta
        ? [
            "-map_metadata", "0",
            "-movflags", "use_metadata_tags",
        ]
        : [];

    const overwrite = program.overwrite ? ["-y"] : [];

    return new Promise((resolve, reject) => {
        const ff = spawn("ffmpeg", [
            "-n",
            "-i", program.input,
            "-c:v", "libx264",
            "-preset", program["v:preset"],
            "-crf", program["v:quality"],
            ...optionalParams,
            "-vsync", "2",
            "-tune", program["v:tune"],
            "-movflags", "+faststart",
            ...meta,
            ...overwrite,
            outputFileName,
        ]);
        console.info(colors.green(ff.spawnargs.join(" ")));
        let lastError = "";
        ff.stdout.on("data", data => console.info(String(data)));
        ff.stderr.on("data", data => {
            lastError = String(data);
            console.error(lastError);
        });
        ff.on("close", (code) => {
            console.info("Process exited with code:", code);
            if (!code) {
                resolve();
                return;
            }
            if (lastError.includes("already exists. Exiting")) {
                resolve();
                return;
            }
            reject(new Error("ffmpeg exited with code: " + code));
        });
    });
};

const getModifiedDate = () => {
    return new Promise((resolve, reject) => {
        execFile("date", ["-R", "-r", program.input], (error, stdout, stderr) => {
            if (error) {
                reject(new Error(error)); // @todo can do better error?
                return;
            }
            resolve(stdout);
        });
    });
};

const setModifiedDate = (file, newDate) => {
    return new Promise((resolve, reject) => {
        execFile("touch", ["-d", newDate, file], (error, stdout, stderr) => {
            if (error) {
                reject(new Error(error)); // @todo can do better error?
                return;
            }
            resolve(stdout);
        });
    });
};

const fixDate = async () => {
    const date = await getModifiedDate(program.input);
    await setModifiedDate(outputFileName, date);
};

const sizeDiff = async () => {
    const oldFile = await fs.stat(program.input);
    const newFile = await fs.stat(outputFileName);
    const diff = oldFile.size - newFile.size;
    console.info("");
    console.info(`Old file size: ${pretty(oldFile.size)}`);
    console.info(`New file size: ${pretty(newFile.size)}`);
    if (diff > 0) {
        console.info("You have saved", pretty(diff));
    }
    else if (diff < 0) {
        console.info("New file is larger by", pretty(-diff));
    }
    else {
        console.info("File sizes are exactly the same! Amazing!");
    }
};

(async () => {
    try {
        await encode();
        await fixDate();
        await sizeDiff();
        console.info("");
        console.info("Success!");
    }
    catch (e) {
        console.error("");
        console.error("Error happened");
        console.error(e);
    }
})();


src/cmd.mjs
/* eslint-disable no-process-exit */

import program from "commander";
import fs from "fs";
import createError from "better-custom-error";
import { join } from "path";

import __dirname from "./dirname.mjs";

const DEFAULT_QUALITY = 25;
const DEFAULT_PRESET = "slow";
const DEFAULT_TUNE = "film";

const ArgumentError = createError("ArgumentError");

// @TODO deal with that ⬇
// eslint-disable-next-line no-sync
const packageJson = JSON.parse(String(fs.readFileSync(join(__dirname, "..", "package.json"))));

const number = (value) => {
    return parseInt(value, 10);
};

const float = (value) => {
    return parseFloat(value);
};

const kbps = (value) => {
    return value + "k";
};

const rotateMap = {
    "90": "transpose=1",
    "180": ["transpose=2", "transpose=2"],
    "270": "transpose=2",
    "-90": "transpose=2",
};

const rotate = (value) => {
    const vf = rotateMap[value];
    if (!vf) {
        throw new ArgumentError(`Rotation ${value} is incorrect`);
    }
    return vf;
};

const filePath = (value) => {
    try {
        fs.accessSync(value); // eslint-disable-line no-sync
        return value;
    }
    catch (e) { // eslint-disable-line no-unused-vars
        throw new ArgumentError(`File ${value} doesn't exist or is inaccessible.`);
    }
};

const cmd = () => { // eslint-disable-line max-statements, max-lines-per-function
    try {
        program.version(packageJson.version, "-v, --version")
            .option("-i, --input <path>", "source file to encode", filePath)
            .option(
                "--v:quality <number>",
                "sets video quality (0-51, where 0 is lossless, 17 is visually lossless, 23 is optimal default)",
                number,
                DEFAULT_QUALITY,
            )
            .option(
                "--v:fps <number>",
                "sets output video framerate",
                float,
            )
            .option(
                "--v:rotate <degrees>",
                "rotates video, only `90`, `180`, `270` are valid values",
                rotate,
            )
            .option(
                "--v:scale <degrees>",
                "scales video, feed it with raw ffmpeg `scale` `vf`, basic example: 1280x720, 1280:-1, -1:720",
            )
            .option(
                "--v:preset <preset>",
                "sets video encoding preset", // @todo list presets as validation
                DEFAULT_PRESET,
            )
            .option(
                "--v:tune <tune>",
                "sets video tune preset", // @todo list presets as validation
                DEFAULT_TUNE,
            )
            .option("--a:copy", "copies source audio")
            .option("--a:none", "disables audio")
            .option("--a:channels <number>", "sets audio channels count", number)
            .option(
                "--a:bitrate <number>",
                "sets audio quality (in kbps)",
                kbps,
            )
            .option(
                "--from <string>",
                "cuts video from this time",
            )
            .option(
                "--to <string>",
                "cuts video to this time",
            )
            .option("--no-meta", "skips metadata copying")
            .option("-y, --overwrite", "overwrites encoded file if already exists")
            .parse(process.argv);
    }
    catch (e) {
        if (e instanceof ArgumentError) {
            console.error("error:", e.message);
            process.exit(1);
        }

        console.error("Unexpected error");
        console.error(e);
        process.exit(1);
    }

    if (!program.input) {
        console.error("error: required option `-i, --input <path>` missing");
        process.exit(1);
    }

    if (program["a:copy"]) {
        if (program["a:bitrate"] || program["a:channels"]) {
            console.error("error: you can't configure audio settings if you set to copy it");
            process.exit(1);
        }
        if (program["a:none"]) {
            console.error("error: you can't disable audio if you set to copy it");
            process.exit(1);
        }
    }

    if (program["a:none"]) {
        if (program["a:bitrate"] || program["a:channels"]) {
            console.error("error: you can't configure audio settings if you disabled");
            process.exit(1);
        }
    }

    return program;
};

export default cmd;
